text/0000-thread-affinity.md

```diff
@@ -0,0 +1,133 @@
+- Feature Name: thread_affinity
+- Start Date: 2016-01-26
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Thread affinity provides functionality to lock/unlock a specific thread to a specified CPU core.
+The locked threads are then not scheduled so they do not ever move to other cores by the OS scheduler. I believe it useful for a systems programmer to have the ability to set/unset set thread affinity for Rust threads.
+
+# Motivation
+[motivation]: #motivation
+
+My motivation derives from a personal use-case of creating a fiber-based server/client application framework for Rust in my spare time. I need the ability to set thread affinity of Rust threads in an OS-indepedent manner to CPU cores to minimize cache misses and thread context switching.
+
+The proposed feature provides a performance optimization for servers. It would also be useful for MIO and similar frameworks where the main event-loop is locked to one core and passes connections to workers that are also locked to the remaining cores so it takes advantage of cache-locality.
+
+# Detailed design
+[design]: #detailed-design
+
+I strongly encourage others to improve my RFC since I am not a Rust expert and only serves to start a dialog on a proposed API.
+
+I split it into two parts: *User API* and *Implementation*.
+
+### Locking to an arbitrary free CPU core
+```
+let t = thread::spawn(move || {
+    let cpu_lock = CpuLock::lock().unwrap(); // the current thread is now locked to an arbitrary free CPU core
+    let cpu_id = cpu_lock.cpu_id();
+    // thread logic goes here
+    // the cpu_lock will unlock from the core once out of scope
+});
+```
+
+Here's the equivalent but with some syntatical sugar to hide the lock if the user doesn't care to access it:
+
+```
+let t = thread::lock(move || {
+    // thread logic goes here
+    // the cpu_lock will unlock from the CPU core once out of scope
+});
+```
+
+The user may want more fine grained control of which core it should be locked to.
+
+### Locking to a specific CPU core
+
+```
+// locks thread to cpu core 2
+let t = thread::spawn(move || {
+    let cpu_no = 2;
+    let cpu_lock = CpuLock::lock_on(cpu_no).unwrap(); // the current thread is now locked to an arbitrary free CPU core
+    // thread logic goes here
+    // the cpu_lock will unlock from the core once out of scope
+});
+```
+
+```
+// locks thread to cpu core 2
+let cpu_no = 2;
+let t = thread::lock_on(2,move || {
+    let cpu_lock = CpuLock::lock_on(cpu_no).unwrap(); // the current thread is now locked to an arbitrary free CPU core
```

Comment:
if `lock_on` is analogous to `lock`, this line and the one after it shouldn't be here